modules/gapi/src/streaming/vpl/vpl_cpu_accel.cpp

```diff
@@ -0,0 +1,221 @@
+                                 "cannot construct surface index: " + std::to_string(i) + ", error: " +
+                                 ex.what() +
+                                 "Requested surface count: " + std::to_string(pool_size) +
+                                 ", surface bytes: " + std::to_string(surface_size_bytes));
+    }
+
+    // remember pool by key
+    GAPI_LOG_INFO(nullptr, "New pool allocated, key: " << preallocated_pool_memory_ptr <<
+                           ", surface count: " << pool.size() <<
+                           ", surface size bytes: " << surface_size_bytes);
+    try {
+        if (!pool_table.emplace(preallocated_pool_memory_ptr, std::move(pool)).second) {
+            throw std::runtime_error(std::string("VPLCPUAccelerationPolicy::create_surface_pool - ") +
+                                     "cannot insert pool, table size: " + std::to_string(pool_table.size()));
+        }
+    } catch (const std::exception&) {
+        throw;
+    }
+
+    return preallocated_pool_memory_ptr;
+}
+
+VPLCPUAccelerationPolicy::surface_weak_ptr_t VPLCPUAccelerationPolicy::get_free_surface(pool_key_t key) const {
+    auto pool_it = pool_table.find(key);
+    if (pool_it == pool_table.end()) {
+        std::stringstream ss;
+        ss << "key is not found: " << key << ", table size: " << pool_table.size();
+        const std::string& str = ss.str();
+        GAPI_LOG_WARNING(nullptr, str);
+        throw std::runtime_error(std::string(__FUNCTION__) + " - " + str);
+    }
+
+    const pool_t& requested_pool = pool_it->second;
+    auto it =
+        std::find_if(requested_pool.begin(), requested_pool.end(),
+                     [](const surface_ptr_t& val) {
+            GAPI_Assert(val && "Pool contains empty surface");
+            return !val->get_locks_count();
+        });
+
+    // Limitation realloc pool might be a future extension
+    if (it == requested_pool.end()) {
+        std::stringstream ss;
+        ss << "cannot get free surface from pool, key: " << key << ", size: " << requested_pool.size();
+        const std::string& str = ss.str();
+        GAPI_LOG_WARNING(nullptr, str);
+        throw std::runtime_error(std::string(__FUNCTION__) + " - " + str);
+    }
+
+    return *it;
+}
+
+size_t VPLCPUAccelerationPolicy::get_free_surface_count(pool_key_t key) const {
+    auto pool_it = pool_table.find(key);
+    if (pool_it == pool_table.end()) {
+        GAPI_LOG_WARNING(nullptr, "key is not found: " << key <<
+                                  ", table size: " << pool_table.size());
+        return 0;
+    }
+
+    const pool_t& requested_pool = pool_it->second;
+    size_t free_surf_count =
+        std::count_if(requested_pool.begin(), requested_pool.end(),
+                     [](const surface_ptr_t& val) {
+            GAPI_Assert(val && "Pool contains empty surface");
+            return !val->get_locks_count();
```

Comment:
Isn't it a critical section?

Reply:
it uses atomic::load() inside:

Current thread increments it and load only (counter++)
Another threads can decrements it only (counter--).
So i believe we find the first free one ( counter == 0)